ansible/awx
# Allow for projects to be GPG-verified against unexpected changes

##### SUMMARY

This PR comprises the work of the Content Signing Working Group over the last several weeks. It includes a number of changes, including:

- the addition of a new default `GPG Public Key` credential type
- a new `Project` model field to make use of this credential type
- changes to the project update playbook and a new action plugin that make use of `ansible-sign` (as a library) for playbook/project signature validation
- necessary changes to add project update tags to make the above function
- UI changes to accommodate the new credential type and make use of it.
- small change to AWX collection to account for the new feature

Hats off to @djyasin @TheRealHaoLiu and @vedaperi for awesome work in making this come together. This PR is the culmination of much learning and great collaboration. It also marks the end of the `feature-project-integrity` branch, as once it is merged, further work will be done against `devel` instead.

To test this PR sufficiently:

- create a test project (possibly in a temporary repository on GitHub). It can be simple such as just an inventory file and a playbook, or more complex.
- use `ansible-sign` to sign the project with a GPG key (there is a [a rundown](https://ansible.github.io/ansible-sign/rundown.html) of how to use `ansible-sign`), and push your changes to the project repository.
- add your GPG public key to AWX as a credential (`gpg --export --armour <your pubkey fingerprint>` will render it).
- add the test project as an AWX project, and enable content signing on the project by selecting your GPG public key in the project edit page.
- sync the project and ensure that the content-verification tasks get run in the project sync job and are successful
- once that is done, modify the project in various ways and ensure that the module **fails** when expected.
  - Example:
    - You can edit a file, without re-signing the project, commit that change, sync the project, make sure it fails
    - Then revert that change so the project is valid again, edit `.ansible-sign/sha256sum.txt` (without re-signing anything), commit that change and ensure validation fails again.

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME

 - API
 - UI
 - Collection
